### PR TITLE
PlatformViewLink, handling creation of the PlatformViewSurface and dispose PlatformViewController

### PR DESCRIPTION
**Pull request by cyanglaz**
## Description

Introduce `PlatformViewLink` which in this PR handles the creation and disposal of the platform view.

## Related Issues

https://github.com/flutter/flutter/issues/36779

## Tests

I added the following tests:

'PlatformViewController Widget init, should create a SizedBox widget before onPlatformViewCreated and a PlatformViewSurface after'
'PlatformViewController Widget dispose'

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 1]
```diff
@@ -581,6 +581,129 @@ class _UiKitPlatformView extends LeafRenderObjectWidget {
...
+}
+
+/// The factory to construct a [PlatformViewSurface].
+///
+/// It is required when constructing a [PlatformViewLink].
+typedef PlatformViewSurfaceFactory = PlatformViewSurface Function(BuildContext context, PlatformViewController controller);
+
+/// Constructs a [PlatformViewController].
+///
+/// The implementer of a new platform view is responsible to implement this method when constructing a [PlatformViewLink].
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]: Constructs a [PlatformViewController].

The [PlatformViewController.id] field of the created controller must match the value of the `params` [PlatformViewCreationParams.id]  field.

See also [PlatformViewLink.onCreate].

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 2]
```diff
@@ -581,6 +581,129 @@ class _UiKitPlatformView extends LeafRenderObjectWidget {
...
+typedef CreatePlatformView = PlatformViewController Function(PlatformViewCreationParams params);
+
+/// The widget responsible for handling the life cycle and focus for a platform view.
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]: Links a platform view with the Flutter framework.

Provides common functionality for embedding a platform view (e.g an `android.view.View` on Android) with the Flutter framework.

{@macro flutter.widgets.platformViews.lifetime}

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 3]
```diff
@@ -581,6 +581,129 @@ class _UiKitPlatformView extends LeafRenderObjectWidget {
...
+
+  /// Construct a `PlatformViewLink` widget.
+  ///
+  /// The [surfaceFactory] and the [createPlatformView] must not be null.
+  ///
+  /// See also:
+  /// * [PlatformViewSurface] for details on the widget returned by `surfaceFactory`.
+  /// * [PlatformViewCreationParams] for how each parameter can be used when implementing `createPlatformView`.
+  const PlatformViewLink({
+    @required PlatformViewSurfaceFactory surfaceFactory,
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]: Add  a key

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 4]
```diff
@@ -581,6 +581,129 @@ class _UiKitPlatformView extends LeafRenderObjectWidget {
...
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _initializeOnce();
+  }
+
+  void _onPlatformViewCreated(int id) {
+     _platformViewCreated = true;
+    setState((){});
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]: setState(() {
  _platformViewCreated = true;
});

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 5]
```diff
@@ -581,6 +581,129 @@ class _UiKitPlatformView extends LeafRenderObjectWidget {
...
+    if (_initialized) {
+      return;
+    }
+    _initialized = true;
+    _id = platformViewsRegistry.getNextPlatformViewId();
+    _controller = widget._createPlatformView(PlatformViewCreationParams._(id:_id, onPlatformViewCreated:_onPlatformViewCreated));
+  }
+
+  @override
+  void didChangeDependencies() {
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]: Can we override `initState` instead?

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 6]
```diff
@@ -581,6 +581,129 @@ class _UiKitPlatformView extends LeafRenderObjectWidget {
...
+///            gestureRecognizers: gestureRecognizers,
+///            controller: controller,
+///            hitTestBehavior: PlatformViewHitTestBehavior.opaque,
+///        );
+///       },
+///    );
+///   }
+/// }
+/// ```
+class PlatformViewLink extends StatefulWidget {
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]: Document what happens when the widget is rebuild with a different create callback or a different surfaceFactory

File: packages/flutter/test/widgets/platform_view_test.dart [hunk 7]
```diff
@@ -1932,5 +1932,59 @@ void main() {
...
+            hitTestBehavior: PlatformViewHitTestBehavior.opaque,
+        );
+      });
+
+      await tester.pumpWidget(platformViewLink);
+
+      await tester.pumpWidget(Container());
+
+      expect(disposedController.disposed, true);
+    });
```
Comment (amirh) [inline packages/flutter/test/widgets/platform_view_test.dart]: I would add an equivalent test to `Android view survives widget tree change`

File: packages/flutter/lib/src/widgets/platform_view.dart (hunk 6 above)
Comment (cyanglaz) [reply packages/flutter/lib/src/widgets/platform_view.dart]: done, with the change that made sure `PlatformViewSurface` only created once.

**Review by cyanglaz — COMMENTED** [review]
Updated with review comments.

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 8]
```diff
@@ -581,6 +581,139 @@ class _UiKitPlatformView extends LeafRenderObjectWidget {
...
+  ///
+  /// Must be invoked as soon as the embedded platform view is ready.
+  /// For example, on iOS, this has to be called when we know the `UIView` is initialized.
+  final PlatformViewCreatedCallback onPlatformViewCreated;
+}
+
+/// The factory to construct a [PlatformViewSurface].
+///
+/// It is required when constructing a [PlatformViewLink].
+typedef PlatformViewSurfaceFactory = PlatformViewSurface Function(BuildContext context, PlatformViewController controller);
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]: We probably want to make this return a widget.

Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]:
```dart
/// A factory for a surface presenting a platform view as part of the widget hierarchy.
///
/// The returned widget should present the platform view associated with `controller`.
///
///  See also:
///   * [PlatformViewSurface], a common widget for presenting platform views.
typedef PlatformViewSurfaceFactory = Widget Function(BuildContext context, PlatformViewController controller);
```

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 9]
```diff
@@ -581,6 +581,139 @@ class _UiKitPlatformView extends LeafRenderObjectWidget {
...
+                                  _surfaceFactory = surfaceFactory,
+                                  _createPlatformViewController = createPlatformViewController,
+                                  super(key: key);
+
+
+  final PlatformViewSurfaceFactory _surfaceFactory;
+  final CreatePlatformViewController _createPlatformViewController;
+
+  @override
+  State<StatefulWidget> createState() {
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]: style nit: use `=>` if it fits in one line.

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 10]
```diff
@@ -581,6 +581,139 @@ class _UiKitPlatformView extends LeafRenderObjectWidget {
...
+  /// For example, on iOS, this has to be called when we know the `UIView` is initialized.
+  final PlatformViewCreatedCallback onPlatformViewCreated;
+}
+
+/// The factory to construct a [PlatformViewSurface].
+///
+/// It is required when constructing a [PlatformViewLink].
+typedef PlatformViewSurfaceFactory = PlatformViewSurface Function(BuildContext context, PlatformViewController controller);
+
+///Constructs a [PlatformViewController].
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/flutter/lib/src/widgets/platform_view.dart]: done

File: packages/flutter/lib/src/widgets/platform_view.dart (hunk 9 above)
Comment (cyanglaz) [reply packages/flutter/lib/src/widgets/platform_view.dart]: done

**Review by cyanglaz — COMMENTED** [review]
Updated. Ready for another review.

**Review by amirh — APPROVED** [review]
LGTM

**Pull request merged**